### PR TITLE
[train] add built-in cross-validation API

### LINKED ISSUE #53581: [Ray Train Feature Request] Native Cross-Validation Support in Ray Train API
### Description

Current workflows for e.g. for XGBoostTrainer etc do not allow for cross validation 
- (https://scikit-learn.org/stable/modules/cross_validation.html) 
- https://xgboost.readthedocs.io/en/latest/python/examples/cross_validation.html

The user has to implement cross validation manually using for loops which arent vectorised and thus take a considerable time to complete.

Also in my experiments, scikit learn models that implement cv arent really well parallelised and take longer as well. Thus this feature request suggest that we implement cross validation within Ray Train API, so can be applied across any framework.

*see use case section in this issue for more details*

### Use case

I want to efficiently perform cross-validation at scale, using distributed compute, as part of my end-to-end machine learning workflows on Ray.

**Context:**
Cross-validation is a core part of robust ML model development. It helps estimate the generalisation ability of models and is a standard practice for evaluating algorithms, tuning hyperparameters, and comparing models. Most ML practitioners expect to use cross-validation as a simple, integrated step in their workflow.
- K-Fold Cross-Validation
- Stratified K-Fold Cross-Validation
- Group K-Fold
- Time Series Split (I can see an argument for this due to its complexity and how each framework deal with this, that perhaps frameworks like Nixtla, Darts implement so this can be off loaded to them (https://nixtlaverse.nixtla.io/mlforecast/docs/how-to-guides/cross_validation.html#cross-validation)) 

**Current Situation:**
When working with Ray’s trainers (such as XGBoostTrainer etc.), there is no built-in support for cross-validation.

As a user, this means I need to write my own loops or orchestrate manual splitting and training, which is both time-consuming and less efficient. Standard libraries (e.g., scikit-learn, XGBoost) do provide cross-validation, but they either do not scale to distributed settings or do not fully utilise cluster resources.

In practice, this makes large-scale cross-validation slow and cumbersome, especially with big datasets or more complex models.

**What I want:**
A way to seamlessly and efficiently apply cross-validation when training models with Ray—leveraging Ray’s distributed capabilities—so that I can:

Use cross-validation as a first-class citizen in my training workflow
- Avoid reimplementing common evaluation logic
- Run cross-validation jobs faster and more efficiently, at any scale

**Ideal outcome:**
When I train a model using Ray, I want cross-validation to be as accessible and scalable as the rest of the API—making it easy to get robust model performance estimates, without extra boilerplate or resource underutilisation.

### PR DESCRIPTION
**Pull request by EmaFerrao**
## Description

Add a built-in cross-validation API for Ray Train to simplify and standardize cross-validation workflows across distributed training jobs.

Currently, users must manually implement cross-validation logic by repeatedly splitting datasets and instantiating trainers for each fold. This leads to boilerplate code and prevents consistent integration with Ray’s distributed execution model.

This PR introduces a unified API composed of:

- `cross_validate`: main entry point that orchestrates training across folds
- `CVResult`: structured output containing per-fold results and aggregated metrics (mean and std)
- `Splitter` (abstract base class): defines the interface for dataset partitioning strategies
- Built-in splitter implementations:
  - `KFoldSplitter`
  - `StratifiedKFoldSplitter`
  - `GroupedKFoldSplitter`
  - `TimeSeriesSplitter`

Each fold is executed sequentially, with a fresh trainer instance per fold. Train and validation datasets are injected into the trainer via the `"train"` and `"val"` dataset keys.

---

## Implementation notes

- Cross-validation folds are executed sequentially to avoid resource contention between distributed trainer runs.
- Dataset materialization:
    - `StratifiedKFoldSplitter` materializes the dataset at an intermediate step to avoid recomputing `groupby + map_groups` for every train/val split per fold.
    - `TimeSeriesSplitter` materializes after sorting by `time_column` when one is provided, to establish temporal order. Without this, re-executing the sort on a dataset with identical timestamps could produce a different ordering.
- Dataset splitting is delegated to pluggable `Splitter` implementations.
- Metric aggregation is performed across folds for all numeric metrics reported in `ray.train.Result`.
- The API is designed to be extensible, allowing the implementation of custom splitters by subclassing `Splitter`.

---

## Evaluation

The feature was benchmarked using a `DummyTrainer` that counts validation rows to isolate the overhead of dataset splitting and orchestration.

We compared `cross_validate` against equivalent manual implementations across different dataset sizes and splitter strategies.

Key observations:

- For small datasets (≤ ~500k rows), manual implementations are faster due to lower overhead (no distributed orchestration).
- Around ~1M rows, both approaches converge in performance.
- For larger datasets (≥ 20M rows), `cross_validate` becomes significantly faster, achieving up to ~4x–18x speedup depending on splitter and workload.

---

## Documentation

User-facing documentation was added under Ray Train User Guides, including:

- Overview of cross-validation API
- Examples for all built-in splitters
- Guidance on sequential execution model and dataset handling

API references were added to `api.rst`.

---

## Related issues

Closes #53581

### CODE REVIEW

File: python/ray/train/cross_validation/kfold/_hashbased.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        # We intentionally keep the pipeline streaming here (no materialize).
+        # Materializing the dataset after adding a fold column would avoid
+        # re-running the `map_batches` work during the 2*n_splits filter
+        # passes performed by `_build_folds_from_column`, but it can cause
+        # excessive memory use for very large datasets.
+        dataset_with_fold_id = dataset.map_batches(
+            add_fold_column, batch_format="pandas"
+        )
+
+        return _build_folds_from_column(dataset_with_fold_id, fold_column, n_splits)
```
[the remark this answers is not in the document]
Comment (EmaFerrao) [reply python/ray/train/cross_validation/kfold/_hashbased.py:115]: Fixing this would require calling count() on each fold before passing it to the trainer, which for hash-based splitters forces premature materialization, going against their intentional streaming design. The empty fold case is also rare enough that we don't think it justifies that cost for all users. We've instead documented this limitation clearly in the docstrings of HasBasedKFoldSplitter

**Comment by EmaFerrao** [comment]
Hi @justinvyu @matthewdeng, any chance you could take a look at this PR? Let us know if anything should be changed, thanks!

**Review by pseudo-rnd-thoughts — COMMENTED** [review]
Thanks for the PR @EmaFerrao 
Reading the solution, it overall looks good however I'm nervous about the number of different solutions added. The challenge is that adding the code is relatively simple but we need to maintain this code going forward, for which I don't think there many requests currently.

How popular is KFold and are you aware of other requests for this in KFold? What would the minimal implementation be?

**Comment by EmaFerrao** [comment]
Thanks for taking a look @pseudo-rnd-thoughts!

K-fold itself has been requested before (e.g. [#33677](https://github.com/ray-project/ray/issues/33677)), and cross-validation more broadly has come up repeatedly in hyperparameter tuning workflows (e.g. [#6560](https://github.com/ray-project/ray/issues/6560), [#7744](https://github.com/ray-project/ray/issues/7744)).

As for the minimal implementation, we’d say it’s `cross_validate` + `Splitter` + `KFoldSplitter`. `Splitter` provides the extension point for custom logic, and KFold is what most users referenced in the stated issues, so it’s worth including as the default implementation.

The main scope question, in our view, is whether we want to ship additional splitter implementations beyond KFold in the same PR. We included a few because they capture guarantees KFold cannot express structurally and were already requested in [#53581](https://github.com/ray-project/ray/issues/53581). That said, both options seem reasonable: either keeping KFold-only for a minimal initial version, or including the additional splitters and accepting the slightly broader scope.
